### PR TITLE
support 3D attention mask in bert

### LINKED ISSUE #31036: sdpa for bert should support 4D  attention mask.
### Feature request

Currently this is only possible with 2D-mask when sdap is enabled.
```python
# modeling bert
        # Expand the attention mask
        if use_sdpa_attention_masks:
            # Expand the attention mask for SDPA.
            # [bsz, seq_len] -> [bsz, 1, seq_len, seq_len]
            if self.config.is_decoder:
                extended_attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
                    attention_mask,
                    input_shape,
                    embedding_output,
                    past_key_values_length,
                )
            else:
                extended_attention_mask = _prepare_4d_attention_mask_for_sdpa(
                    attention_mask, embedding_output.dtype, tgt_len=seq_length
                )
            #  print(f"sdpa attn mask:{embedding_output.dtype=} {extended_attention_mask.sum(axis=(2,3))=}")
        else:
            # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
            # ourselves in which case we just need to make it broadcastable to all heads.
            extended_attention_mask = self.get_extended_attention_mask(attention_mask, input_shape)
```

### Motivation

4D attention mask is widely used such as padding bert.

We construct 4D attention mask before the model forward and pass it to the forward call.

I think it is easy to exend the current implement of _expand_mask to support 3D/4D attention mask.

```python
    @staticmethod
    def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
        """
        Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
        """
        bsz, *_, src_len = mask.size()
        tgt_len = tgt_len if tgt_len is not None else src_len

        if mask.ndim == 2:
            expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
        elif mask.ndim == 3:
            expanded_mask = mask.unsqueeze(1).to(dtype)
        elif mask.ndim == 4:
            expanded_mask = mask.to(dtype)

        inverted_mask = 1.0 - expanded_mask
        #  print(f"{dtype=} {torch.finfo(dtype).min=} {expanded_mask.bool().sum((1,2,3))=}")

        return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
```

### Your contribution

I can do PR and contribution.

### PR DESCRIPTION
**Pull request by gathierry**
# What does this PR do?

Fixes #31036
Try to fix with minimal change as mentioned in #31302

3D attention masks are used in [GroundingDINO](https://github.com/IDEA-Research/GroundingDINO)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- text models: @ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Cool! Let's document that as well and good to merge 🤗 (I think you might have some make fix copied to handle)

**Comment by gathierry** [comment]
Hi I updated the doc string. But I'm not familiar with how the `attention_mask` doc is overwritten. If I did it in a wrong way, could you help indicate where should be change? 
`make fix-copies` seems fine, but maybe I miss something.

**Comment by gathierry** [comment]
Hi @ArthurZucker , could you give another review on the updated code? Thanks

**Review by ArthurZucker — COMMENTED** [review]
LGTM, I think this is more `support arbitrary mask shape`, as if the shape is not == 2 we are just not expanding the mask! 

The mask will be broadcasted to 4d eventually, which is why I would rather make sure our inputs are 4d or 2d TBH! But otherwise LGTM

**Comment by gathierry** [comment]
Thanks for the comment @ArthurZucker. The use case is that Grounding DINO uses a 3D attention mask. Right now, all implementations have to creating a wrapper of `BertModel`. Accepting 3D mask is more straightforward but I'm also okay with the idea about only accepting 2d or 4d. Grounding DINO has to expand its attention mask beforehand though.
Do you want an `assert` in the beginning to make sure the input shape is either 2D or 4D?

**Comment by gathierry** [comment]
Could you give a reference of the check in Llama, I can try to do the same

**Comment by ArthurZucker** [comment]
MB, in https://github.com/huggingface/transformers/blob/834ec7b1ccf9b82905d5803aca41e1744849dae7/src/transformers/models/llama/modeling_llama.py#L59-L109 we don't raise and error so let's go with that

**Comment by gathierry** [comment]
For BERT it will go to some common functions which makes it more like the PR #31302

File: src/transformers/models/bert/modeling_bert.py [hunk 1]
```diff
@@ -908,7 +908,7 @@ class BertForPreTrainingOutput(ModelOutput):
             [`PreTrainedTokenizer.__call__`] for details.
 
             [What are input IDs?](../glossary#input-ids)
-        attention_mask (`torch.FloatTensor` of shape `({0})`, *optional*):
+        attention_mask (`torch.FloatTensor` of shape `({0})`or `(batch_size, sequence_length, target_length)`, *optional*):
```
Comment (ArthurZucker) [inline src/transformers/models/bert/modeling_bert.py:911]: let's maybe recommend 4d here!

**Review by ArthurZucker — COMMENTED** [review]
Ok let's just recommend 4d over 3d

**Comment by gathierry** [comment]
I just realize that if we use 4D attention mask, there will be an error at https://github.com/gathierry/transformers/blob/97903a6352fdc897e13bc4f221fc57aa73e2697b/src/transformers/models/bert/modeling_bert.py#L1113
because the function `get_extended_attention_mask` only accept 2D and 3D inputs https://github.com/gathierry/transformers/blob/51296712900f28023cecea6ff7333af4b3c81d30/src/transformers/modeling_utils.py#L1132

**Comment by ArthurZucker** [comment]
yeah we should not got down that code path. 4d mask should be untouched

**Comment by gathierry** [comment]
Yeah but seems like 3D is acceptable?

**Comment by ArthurZucker** [comment]
Okay, let's go with 3d then 😉

**Review by ArthurZucker — APPROVED** [review]
Ready to merge?

**Comment by gathierry** [comment]
Yes, thanks

**Comment by ArthurZucker** [comment]
thanks for your contribution! 🤗
